### ISSUE TITLE
kube-proxy would not update iptables rules when pod selectors are used for service

### ISSUE THREAD

**Issue by ipatrykx** [kubernetes/kubernetes#98066]
**What happened**:

After deploying pod and service with selector and changing the endpoint using kubectl apply, the endpoint's IP reported by `kubectl get ep` is updated to new value but iptables rules are not being updated. The same operation works if service is deployed without selector.

**What you expected to happen**:

After applying new definition of endpoint I would expect that IP address reported by `kubectl get ep` and iptables rules would be updated to newly applied values.

**How to reproduce it (as minimally and precisely as possible)**:

Simple nginx pod and service with app selector are applied:

```
apiVersion: v1
kind: Pod
metadata:
  name: nginx-pod
  labels:
    app: nginx
spec:
  containers:
  - image: nginx
    imagePullPolicy: IfNotPresent
    name: nginx
  restartPolicy: Always
```

```
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  selector:
    app: nginx
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9999
```

The pod, service and endpoint is successfully created:

```
[root@virtual1 ~]# kubectl get pod -o wide
NAME        READY   STATUS    RESTARTS   AGE   IP             NODE       NOMINATED NODE   READINESS GATES
nginx-pod   1/1     Running   0          8m    10.244.0.160   virtual1   <none>           <none>

[root@virtual1 ~]# kubectl get svc
NAME            TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)   AGE
kubernetes      ClusterIP   10.96.0.1        <none>        443/TCP   14d
nginx-service   ClusterIP   10.111.147.147   <none>        80/TCP    9m32s

[root@virtual1 ~]# kubectl get ep
NAME            ENDPOINTS             AGE
kubernetes      192.168.122.58:6443   14d
nginx-service   10.244.0.160:9999     9m33s
```

iptables rules are also created by kube-proxy:

```
[root@virtual1 ~]# iptables-save | grep nginx-service
-A KUBE-SEP-FXSJGQ4NBPW72IH6 -s 10.244.0.160/32 -m comment --comment "default/nginx-service" -j KUBE-MARK-MASQ
-A KUBE-SEP-FXSJGQ4NBPW72IH6 -p tcp -m comment --comment "default/nginx-service" -m tcp -j DNAT --to-destination **10.244.0.160**:9999
-A KUBE-SERVICES ! -s 10.244.0.0/16 -d 10.111.147.147/32 -p tcp -m comment --comment "default/nginx-service cluster IP" -m tcp --dport 80 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.111.147.147/32 -p tcp -m comment --comment "default/nginx-service cluster IP" -m tcp --dport 80 -j KUBE-SVC-V2OKYYMBY3REGZOG
-A KUBE-SVC-V2OKYYMBY3REGZOG -m comment --comment "default/nginx-service" -j KUBE-SEP-FXSJGQ4NBPW72IH6
```

Now I would like to change the endpoint's address, so I apply endpoint with the same same name as already existing endpoint but with different address (192.0.42.101 instead of 10.244.0.160 in my example):

```
apiVersion: v1
kind: Endpoints
metadata:
  name: nginx-service
subsets:
  - addresses:
      - ip: 192.0.42.101
    ports:
      - port: 9999
```

Right after that I check with `kubectl get ep` if the IP was updated, and it is:

```[root@virtual1 ~]# kubectl get ep
NAME            ENDPOINTS             AGE
kubernetes      192.168.122.58:6443   14d
nginx-service   192.0.42.101:9999     23m
```

However iptables still uses the same address as before (10.244.0.160 here):

```
[root@virtual1 ~]# iptables-save | grep nginx-service
-A KUBE-SEP-FXSJGQ4NBPW72IH6 -s 10.244.0.160/32 -m comment --comment "default/nginx-service" -j KUBE-MARK-MASQ
-A KUBE-SEP-FXSJGQ4NBPW72IH6 -p tcp -m comment --comment "default/nginx-service" -m tcp -j DNAT --to-destination 10.244.0.160:9999
-A KUBE-SERVICES ! -s 10.244.0.0/16 -d 10.111.147.147/32 -p tcp -m comment --comment "default/nginx-service cluster IP" -m tcp --dport 80 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.111.147.147/32 -p tcp -m comment --comment "default/nginx-service cluster IP" -m tcp --dport 80 -j KUBE-SVC-V2OKYYMBY3REGZOG
-A KUBE-SVC-V2OKYYMBY3REGZOG -m comment --comment "default/nginx-service" -j KUBE-SEP-FXSJGQ4NBPW72IH6
```

**Anything else we need to know?**:

What I've tried to do, works perfectly well when service is not using selectors, eg, applying:

```
apiVersion: v1
kind: Service
metadata:
  name: test-service
spec:
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9999

apiVersion: v1
kind: Endpoints
metadata:
  name: test-service
subsets:
  - addresses:
      - ip: 192.0.42.101
    ports:
      - port: 9999

 ```

will result in following iptables:

```
[root@virtual1 ~]# iptables-save | grep test-service
-A KUBE-SEP-UDFPMNWDCHTE2OML -s 192.0.42.101/32 -m comment --comment "default/test-service" -j KUBE-MARK-MASQ
-A KUBE-SEP-UDFPMNWDCHTE2OML -p tcp -m comment --comment "default/test-service" -m tcp -j DNAT --to-destination 192.0.42.101:9999
-A KUBE-SERVICES ! -s 10.244.0.0/16 -d 10.98.159.247/32 -p tcp -m comment --comment "default/test-service cluster IP" -m tcp --dport 80 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.98.159.247/32 -p tcp -m comment --comment "default/test-service cluster IP" -m tcp --dport 80 -j KUBE-SVC-AEJ565DT7YN2S3PD
-A KUBE-SVC-AEJ565DT7YN2S3PD -m comment --comment "default/test-service" -j KUBE-SEP-UDFPMNWDCHTE2OML
[root@virtual1 ~]# 
```

Applying new endpoint on top of that:

```
apiVersion: v1
kind: Endpoints
metadata:
  name: test-service
  annotations:
    endpointType: "secondary"
subsets:
  - addresses:
      - ip: 200.0.0.200
    ports:
      - port: 9999
```

results in updated iptables:

```
[root@virtual1 ~]# iptables-save | grep test-service
-A KUBE-SEP-CR74YFSMMWL6MICX -s 200.0.0.200/32 -m comment --comment "default/test-service" -j KUBE-MARK-MASQ
-A KUBE-SEP-CR74YFSMMWL6MICX -p tcp -m comment --comment "default/test-service" -m tcp -j DNAT --to-destination 200.0.0.200:9999
-A KUBE-SERVICES ! -s 10.244.0.0/16 -d 10.98.159.247/32 -p tcp -m comment --comment "default/test-service cluster IP" -m tcp --dport 80 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.98.159.247/32 -p tcp -m comment --comment "default/test-service cluster IP" -m tcp --dport 80 -j KUBE-SVC-AEJ565DT7YN2S3PD
-A KUBE-SVC-AEJ565DT7YN2S3PD -m comment --comment "default/test-service" -j KUBE-SEP-CR74YFSMMWL6MICX
```

**Environment**:
- Kubernetes version (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-18T12:09:25Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-18T12:00:47Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration: n/a
- OS (e.g: `cat /etc/os-release`): CentOS Linux release 7.6.1810 (Core) 
- Kernel (e.g. `uname -a`): Linux virtual1 3.10.0-957.el7.x86_64 #1 SMP Thu Nov 8 23:39:32 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: kubeadm
- Network plugin and version (if this is a network-related bug): n/a
- Others: n/a

**Comment by aojea** [comment]
hmm I've verified that you can mutate a generated endpoint from a service with a selector, this looks like a bug to me, what do you think @thockin ?

**Comment by thockin** [comment]
Being able to modify endpoints is not a 'bug', in the sense that we can or should fix it.  The user has permission to edit Endpoints, so they can edit it.  That doesn't mean the controller won't blow it away  at any point thereafter.

I'm guessing here, but probably kube-proxy is watching EndpointSlice, rather than Endpoints, so it never sees this change.  The reason it works for selectorless Services is that we have a different controller (endpoint mirroring), which syncs naked Endpoints into EndpointSlices when it is not auto-managed (for exactly this case).

TL;DR - working as intended, 96.4% confidence.

**Comment by thockin** [comment]
If you agree, please close?

**Comment by ipatrykx** [comment]
@thockin OK, I get how it works. Thanks for the detailed explanation, that's really helpful!

Could you please tell me if you are maybe aware if there is any way I could update the endpoint address in the corresponding endpointslice? Simply trying to apply yaml on top of the existing endpointslice seems not to work (or I've done something wrong).

**Comment by aojea** [comment]
> Being able to modify endpoints is not a 'bug', in the sense that we can or should fix it.

yeah, bad wording, I tried to mean that I expect the endpoint controller to have the  endpoint object in sync with the "service-selector. podIPs"

> That doesn't mean the controller won't blow it away at any point thereafter.

the endpointslices  controller detects the updates and reconcile the object to have the "right" addresses, the ones that match the PodIPs associated to the service with the selector, @robscott correct me if I´m wrong please

**Comment by robscott** [comment]
@aojea you're completely right. The EndpointSlice controller does have a mechanism built in to detect external changes to EndpointSlices that are managed by the controller. If that happens, the controller will overwrite them. With EndpointSlices there's a very strong concept of which component an EndpointSlice should be managed by. This is accomplished with the `endpointslice.kubernetes.io/managed-by` [label](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/discovery/v1beta1/well_known_labels.go#L27). Both the EndpointSlice and EndpointSlice mirroring controllers will ensure that slices managed by them reflect the controller's desired state. 

You can always choose to add additional slices for a Service by choosing a different value for the `managed-by` label. Controllers only modify slices with a matching value for that label. Alternatively, you can set the Service selector to empty and both the Endpoints and EndpointSlice controller will not create/modify endpoint(slice)s for that Service.

**Comment by thockin** [comment]
@ipatrykx - why do you want to manually edit endpoints that are explicitly managed by a controller?  At best it will eventually get reverted.  At worst (what you see with slices) it will immediately be reverted.  @aojea is proposing to make Endpoints more like EndpointSlices in this way.  

If you want to manually assign endpoints, you can - just remove the selector from your service.

**Comment by aojea** [comment]
I see 2 possible solutions, label the Service with
https://github.com/kubernetes/kubernetes/blob/9f2f1b8c460b4b5ec8bfcd34490fb259ebad65b0/pkg/proxy/apis/well_known_labels.go#L20-L22

or set ClusterIP to None

**Comment by ipatrykx** [comment]
@aojea Thank you, well known label seems to work :)
